### PR TITLE
[SD-1408] Separate data from display form in form fields

### PR DESCRIPTION
**Pull request by jonsterling**
This is a little bit ugly, but it seems like it will suffice; I'm definitely open to any suggestions for making it cleaner.

### CODE REVIEW

File: src/Text/Markdown/SlamDown/Syntax/FormField.purs [hunk 1]
```diff
@@ -11,11 +16,43 @@ import Data.Maybe as M
 import Test.StrongCheck as SC
 import Test.StrongCheck.Gen as SC
 
+-- | Pairing an underlying value and a display form.
+type StringReprR a =
+  { value :: a
+  , display :: String
+  }
+
+newtype StringRepr a = StringRepr (StringReprR a)
```
Comment (garyb) [inline src/Text/Markdown/SlamDown/Syntax/FormField.purs]: Maybe we should call it something like `Option`/`OptionRepr` instead?

Comment (jonsterling) [inline src/Text/Markdown/SlamDown/Syntax/FormField.purs]: Good point, since it's only being used for options...

Comment (garyb) [inline src/Text/Markdown/SlamDown/Syntax/FormField.purs]: `value` and `label` would be the HTML terms for these too, although I'm totally fine with `display` too.

Comment (jonsterling) [inline src/Text/Markdown/SlamDown/Syntax/FormField.purs]: I'll go ahead and call it `label` then, since that may make it more clear what is being done.

**Comment by garyb** [comment]
Looks great :corn:

**Comment by jonsterling** [comment]
@garyb great, thanks! And thanks for the naming suggestions—I think it's a bit more clear what's going on now.

**Comment by garyb** [comment]
Oops, I guess you're not done yet :smile:

**Comment by garyb** [comment]
Ah, just those minor name changes and new getter. :shipit:

**Comment by jonsterling** [comment]
@garyb I also wanted to mention, there is one possible design which I didn't pursue because I was worried it would make the API too gnarly, but maybe it is worth considering.

We could parameterize all `SlamDown, Block, Inline, FormField` by a type `o` which would serve as the underlying values of options—and we could use a type class to provide labels to these. That would allow things to be a little bit more modular, but at the expense of making things more abstruse.

Do you have any thoughts?

**Comment by garyb** [comment]
That had occurred to me too, but I figured we'd end up with an `Either String o` in all those slots, as we'd need to support the normal markdown-syntax-provided options as well as those produced by `eval`. On the upside we could avoid the need to parse the `value` again later, as we could use the index of a selected option to retrieve the `o` value for a selection each time instead.

Now I write it down that does seem like a nice property to have, do you think we should perhaps explore that option?

**Comment by jonsterling** [comment]
@garyb Yeah, the more I think about it, it might be a reasonable way to go about it.

One option for avoiding the `Either` thing by incorporating a certain language basis signature into the type class:

```
class OptionRepr o where
  stringOption ::  String -> o

  optionLabel :: o -> String
```

I'm not sure which is better; what do you think?

**Comment by garyb** [comment]
That definitely works too, it might mean `o ~ Either String o'` in some cases if there is no `o'` that can sensibly constructed from arbitrary strings, but that seems fine to me. It does make the interface a little more "tricky" perhaps, but it does handle the "I don't care about `eval` case" rather nicely as we could provide `SlamDown` as just `type SlamDown = SlamDownP String` rather than invoking `Void` or something.

I'm not suggesting we do call it `SlamDownP` however, I think the `P` scheme has confused enough people already, I just didn't have another name to hand :smile:

**Comment by jonsterling** [comment]
@garyb yeah, I was thinking that it devolves gracefully in case the user doesn't want anything special... I'll try it out and see how this feels.

**Comment by jonsterling** [comment]
All right, I've gone and got rid of the `OptionRepr` stuff and replaced it with parameterization and this `Value` type class.

By the way, if we're still looking for examples where real modules are a big win over their encodings, this probably would count for one.

**Comment by garyb** [comment]
Should we add some `Functor` instances for the types now they all have the value parameter?

**Comment by jonsterling** [comment]
@garyb That's a good idea...

**Comment by garyb** [comment]
Sorry, one last thing: do you think we should export `SlamDown` as a synonym with `a ~ String`? I would have though that most people won't be using the `eval` functionality, so perhaps providing the default that way would be more friendly?

**Comment by jonsterling** [comment]
@garyb It'd indeed be more friendly... Do you have an idea for what we should call the existing `SlamDown`? (Please not `SlamDownP`, :stuck_out_tongue_winking_eye:) 

With that said, I think it is better to give the "good" name to the canonical/basic form of a thing, and a secondary-sounding name to a special-case version of a thing—but this isn't the convention we have followed, so I won't really push for it. (But in my ML libraries, I tend to use things like `X` for the main version, and `SimpleX` for the special case.)

**Comment by garyb** [comment]
Yeah, generally we've used the reverse of that, instead favouring the case that seems more common even if it is a special case in the technical sense. I can see the logic in naming things either way so don't mind how we do it particularly.

Especially as I can indeed not think of a good name :laughing:. `SlamDownF`? :see_no_evil:

**Comment by jonsterling** [comment]
Well, `SlamDownF` isn't terrible, but it makes me think that it is a signature functor that is meant to have its fixed point taken...

**Comment by garyb** [comment]
Yeah same. On that note actually, would we benefit from using `Mu` in here now? We could perhaps do away with the explicit traversals.

**Comment by jonsterling** [comment]
Hmm, it is probably possible to reformulate this to use `Mu`, yeah. Though the traversals we have got are pretty specialized and the change would probably involve rewriting a lot of code...

**Comment by garyb** [comment]
Yeah, maybe we've done enough in here for now! It's been on my mind for a while that I'd like to replace any AST-like things we have with a F+`Mu` implementation, so it can definitely wait a while longer as there's no immediate benefit to changing right now.

Ok, since we haven't really got a good name for the generic `SlamDown`, what would we call the other version? `SlamDownBasic`?

**Comment by jonsterling** [comment]
@garyb I'm fine with `SlamDownBasic` or `SimpleSlamDown`...  Well, thinking about it again, I'm not *super* opposed to calling the one with the type parameter `SlamDownP`—it may confuse people, but most people won't be using that version anyway. 

I'll leave it up to you :smile:

**Comment by garyb** [comment]
I consulted a coin, it said `SlamDownP` 2/3 tosses. :laughing:

**Comment by jonsterling** [comment]
haha, fair enough! :money_with_wings:

**Comment by garyb** [comment]
Once travis completes I think we're finally there :wink:

**Pull request merged**